Updated exception types

chria opened the pull request.

agnesramle commented on src/pyfmi/fmi.pyx:
```diff
@@ -4087,26 +4110,26 @@ cdef class FMUModelBase2(ModelBase):
                 self._fmu_kind = FMIL.fmi2_fmu_kind_cs
             else:
-                raise FMUException("FMUModelBase2 cannot be used directly, use FMUModelME2 or FMUModelCS2.")
+                raise InvalidVersionException("FMUModelBase2 cannot be used directly, use FMUModelME2 or FMUModelCS2.")
```
I'm not convinced this should be an InvalidVersionException, version is 2 in all three cases here.

chria replied:
I agree, it is not ideal. Any suggestion?

agnesramle replied:
For now, I suggest to keep the FMUException.

agnesramle commented on src/pyfmi/fmi.pyx:
```diff
@@ -4115,7 +4138,7 @@ cdef class FMUModelBase2(ModelBase):
             self._modelId           = decode(FMIL.fmi2_import_get_model_identifier_CS(self._fmu))
         else:
-            raise FMUException("FMUModelBase2 cannot be used directly, use FMUModelME2 or FMUModelCS2.")
+            raise InvalidFMUException("FMUModelBase2 cannot be used directly, use FMUModelME2 or FMUModelCS2.")
```
Same as above.

agnesramle commented on src/pyfmi/fmi.pyx:
```diff
@@ -6969,10 +6992,10 @@ cdef class FMUModelCS2(FMUModelBase2):
 
         if self.get_capability_flags()['needsExecutionTool'] == True:
-            raise FMUException('Models that need an execution tool are not supported')
+            raise InvalidVersionException('Models that need an execution tool are not supported')
```
Why should this be an InvalidVersionException? needsExecutionTool has nothing to do with version.

agnesramle commented on src/pyfmi/fmi.pyx:
```diff
@@ -7598,10 +7621,10 @@ cdef class FMUModelME2(FMUModelBase2):
 
         if self.get_capability_flags()['needsExecutionTool'] == True:
-            raise FMUException('Models that need an execution tool are not supported')
+            raise InvalidVersionException('Models that need an execution tool are not supported')
```
Same as above.

agnesramle reviewed (changes requested):
I miss updates of the tests to catch exceptions of the 'child' type and not the 'parent' type. I see it as important since that will also make sure that the exception types are possible to import.
I think there is at least one update in master.pyx needed as well:
raise fmi.FMUException("The Master algorithm currently only supports CS 2.0 FMUs.")
should be InvalidVersionException I think.
The same goes for
raise fmi.FMUException("The coupled model currently only supports ME 2.0 FMUs.")
in fmi_coupled.pyx

chria commented:
Updated after review. We don't have that many tests that catch specific types (and not that many regarding loading exceptions). Updated the one I found.

agnesramle commented:
That we don't have tests regarding loading exceptions is bad. I suggest to add at least one test per exception type here. That will at least test that it is possible to import and catch these Exceptions.

chria commented:
Updated and added a few tests.

agnesramle commented:
LGTM!
